Group pages for units

huss opened the pull request:
# Description

This is a large PR that updates group pages:

- They now work with units.
- They use React hooks.
- They use the new card look.
- Many unused files and functions from the old group pages were removed.

During this work, other changes were rolled in. The larger ones that can be remembered are:

- The style was made consistent on all modal pages and common CSS used.
- Added red star to required items on input on all modal pages.
- Changed input of numbers on modal pages to make easier to do manually and added checks for a number of input values.
- "no unit" is now translated in menus.

Because the entire way groups are edited was updated, this Fixes #413 by eliminating the issue.

I want to thank several developers who worked on parts of this code. Due to unintended issues, some commits lost their name but it is included in the commit message:

- @truongdd03 for the functions to determine the impact of group changes.
- The team consisting of @OsmanSafi, @Alvarez-Jose, @rebecaaaaa & @acuev who helped with the initial versions including the switch to cards.

## Type of change

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

The code has been tested but I expect to do more so some changes could be made. However, it is time to get this important PR for v1.0 out for others to look at and use.

huss committed:
group create page copied

huss committed:
update new group pages

- Create page is in okay state but still has TODOs and needs work.
- Details page only working for create button. Some edits but needs lots
of work.
- View had limited edits but starts from unit.
- Moved two components to hold directory since had name conflict.

huss committed:
First shot group card pages

huss committed:
move old group files, add edit modal page

huss committed:
incomplete edits to group edit page +

OsmanSafi committed:
edits to EditGroupModalComponent

OsmanSafi committed:
Merge branch 'OpenEnergyDashboard:development' into groupPages

huss committed:
fix several group page issues

WIP. Still has issues.

huss committed:
Merge branch 'groupPages' of https://github.com/OsmanSafi/OED into groupPages

huss committed:
create group route working

Also removed old meter translation

huss committed:
old edit session

huss committed:
add missing translation

also fixed minor ordering in file

huss committed:
Merge remote-tracking branch 'origin/development' into groupPages

huss committed:
Merge remote-tracking branch 'upstream/development' into groupPages

huss committed:
simple edit (mostly comments) after rereading

huss committed:
add non-admin group help tooltips

huss committed:
made state cleaner

huss committed:
fix title on group edit

huss committed:
create group button not see by non-admin

huss committed:
fix edit if not admin

The name is show instead of a input text box.
The footer button is close rather than discard/save.

huss committed:
first cut at showing child meters of group

huss committed:
display included groups on edit page

huss committed:
group deep meters displayed on edit

huss committed:
move duplicated functions to util file

huss committed:
move null input check to util function

It is also now used in edit group to avoid console warning

huss committed:
group edit route works

It still has issues. Groups edit made closer to meter
edit. Few other minor edits.

huss committed:
group edit now updates state

- A new type to mirror meters was created for group info.
There are likely a number of actions/reducers that are no
longer needed and will be addressed. It may also be possible
to simplify the group types.
- To reduce the state passed, group edit now only passes the id
and the group info is found from state already being accessed.
- The access of state in the group edit was protected to avoid
a bug when groups were edited.

huss committed:
clean up old actions

The old code tracked group changes in state but the new code does not.
Thus, the actions, reducers & redux types were removed.
It is hoped none will be needed for new code but will see.
The state definitions have not been updated.

huss committed:
comment delete group items

They have TODO since might need later.

huss committed:
make potential graphic unit into util function

huss committed:
groups do fetch once as other types

Added check to group pages in case not done for same reasons
as meters.

huss committed:
dropdown for default graphic unit

Does not currently filter choices

huss committed:
fix linting issues

huss committed:
default graphic unit corrected for non-admin, fixes

huss committed:
default graphic unit menu now dynamic

huss committed:
can edit/save child meters

The look was changed to match the look of the other choice menus.
Some work still needed but seems to work correctly.

huss committed:
merge @truongdd03 work on group functions

huss committed:
add one lost change from merge

huss committed:
code from @truongdd03

I apologize to @truongdd03 for manually copying code.
There were sections that we not wanted so did this to make it faster.
All code from this commit is from @truongdd03.

huss committed:
get all group children change

There are new actions to get all groups child meters/groups to replace
one that did it one group at a time. This removes some issues
with missing state and is more efficient since all group modal pages
created at once. The old actions are left for now.

huss committed:
WIP meter select now filtering initially

huss committed:
fix getting children of group

There was a bug that if the group state was not yet loaded then
getting the children failed. Also, getting the group state currently
overwrites the children arrays.

huss committed:
get to work if non-admin +

The beginning of keeping the code separated for admin/non
is done. Needs more work in this area but the code now work.

Also fixed so space before graphic unit if non-admin.

huss committed:
basic meter selection menu

The possible meters are displayed. They do not update on changes.

huss committed:
editing group meters is mostly working

Change so state works properly for child meters of a group.
Updates the deep meter state when child meters are changed.

huss committed:
Basic graphic unit menu working

huss committed:
Merge remote-tracking branch 'upstream/development' into groupPages

huss committed:
fix discard bug

The code was not properly resetting all state on discard. Also, renamed
variable to be more consistent and differ from an object key.

huss committed:
deep meters correct if not admin

Fix issue where deep meters where not set for non-admin.
Also, fixed sorting of meter children for non-admins.

huss committed:
add notification if group edit success

huss committed:
fix React issue

The code was incorrectly calling a hook (useEffect) inside an if
statement.
See https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level.
Instead the if is placed inside the hook to protect all the code.
Also, the if for the handle save was removed since it was not important.

huss committed:
fix linting issues

huss committed:
fix easy TODOs in edit

huss committed:
remove group outdated

Not planning to use so gone. Also, fixed logic for when should fetch
a group.

huss committed:
one more TODO removed

huss committed:
deep meters correct on edit save

The deep meters of a group needed to be sent to the action so it could
be used to update the group state. It is not sent to the group edit
route.

huss committed:
group selection start

The group selection menu works. There is no update of the deep meters.

huss committed:
remove unused parameter

huss committed:
change useEffect dependency

The change in how child meters/groups are done hopefully means
some of the triggers can be removed;

huss committed:
edit now working except for checks on save

- Have not done careful checks yet or sweep of code.
- Code and design document were updated because you cannot just
diff the meters/groups due to included groups. Deep meters
should now work correctly.
- child meter/group state is updated on each menu change so no longer
done when save.
- fix linting issues.

huss committed:
add no unit as default graphic option

huss committed:
exclude self from group menu

huss committed:
group menu sorted by name

huss committed:
fix tooltips

This applies same fix as PR #863 so they work after modal.

huss committed:
only sort child array when save

Sorting is no longer needed for menus so delay in case not needed.

huss committed:
code check - many changes

- Many small edits to group edit (what mostly checked).
- Translated more strings. This required changes to do "no unit" and
was rolled into meters.
- Made useEffect dependent on child meters/groups.
- Change ListDisplay to remove size.
- Lots of commenting changes.

huss committed:
create group with all edit group items

huss committed:
Group delete + more

- Group delete is basically working.
- Edit/delete is not checking for impacts on other groups and it reloads
the window in all cases. Needs be be made better.
- Fixed but where state was not properly updated on group create.
- Fixed reducer on group details so it does not wipe children if already
there. The code could call both actions where either finishes first.
- Fixed issue where a menu function looked at state without verifying
it was available.
- Received group details now is typed better.

huss committed:
remove deepMeter copy not needed

huss committed:
stops creating empty groups

- Also use TS logic for missing property in state.

huss committed:
only admin can see non-displayable group

huss committed:
Hide child groups if hidden

This does not show child groups if that group is hidden. I left a TODO
because the Redux state has the child group name if hidden and this
should be changed.

huss committed:
fix linting

huss committed:
WIP

huss committed:
created group now in edit

Fixed bug where group created was not available for edit.

huss committed:
fix bug when group children missing on create

The group children state is now fetched when a new group is created.
It also needed to have the group details page depend on this status.
Before this change, a new group did not have the children on edit.

truongdd03 committed:
Merge branch 'groupPages' of https://github.com/huss/OED into groupPages

truongdd03 committed:
change color of LostCompatibleUnits and display color for multiselect's options

huss committed:
translate warning messages

huss committed:
start cross group edit checks

This is not complete but is working at a reasonable level.
The state for edits was changed and functions were moved into component
so it could access state.

- Remove unused group actions/reducers that used to track group edits.

huss committed:
simplify state/logic for display

The code used to store state for the menus/lists displayed. Now they are
directly calculated from the state when a render occurs. This is
simpler and it avoids them getting out of sync.

huss committed:
add/remove meter/group working okay

This code probably needs some work but you can now add/remove a meter
or group. It checks for bad conditions, warns the admin and undoes as
needed. Some other fixes including linting.

huss committed:
disallow delete if in another group

huss committed:
Merge remote-tracking branch 'upstream/development' into groupPages

huss committed:
partly fix group routes to use success/failure

huss committed:
consistent id +

use groupState.id everywhere vs props.groupId. Also, remove unneeded
save check for no children since cannot do any more.

huss committed:
clean up TODOs that not needed.

huss committed:
consistent cancel message

huss committed:
edits from reading again

This is mostly comments. I also removed unneed prop and
renamed some items.

huss committed:
commit updates closer to edit

huss committed:
fix group menu choice bug

Was passing the wrong value. Also, small edits.

huss committed:
style & input checks for all modals

The changes conversions, groups, meters & units modals to:

- Use standard CSS and style imported from a file.
- Make inputs easier.
- Check for invalid inputs in case manually enter.
- Puts red star after required input items.
- Removed all placeholders in input boxes.
- Change some window.alert to nortifyUser.
- Misc. minor edits including removing unneed <br />

huss committed:
Merge pull request #1 from truongdd03/groupPages

Displaying color for options

huss committed:
fix default graphic unit with create

The default graphic unit is updated to no unit if the chosen group/meter
requires this. This is now consistent with edit.

huss committed:
update dependencies on useEffect

Hopefully only the needed dependencies are listed. Removed child
meters/groups since deep meters should update and is what is actually
needed. All now on pik since needed to be loaded and can miss on reload.
Similarly for meters state.
Removed editGroup state since overly broad. So far no issues seen.

huss committed:
only use notifyUser in group code

At some point we probably should change the entire code base and
decide on how all alerts are done (maybe in React way).

huss committed:
remove commented out routes that no longer needed

spearec commented on src/client/app/actions/groups.ts:
```diff
@@ -4,19 +4,18 @@
 import * as t from '../types/redux/groups';
 import { groupsApi } from '../utils/api';
-import { browserHistory } from '../utils/history';
+// TODO import { browserHistory } from '../utils/history';
```
unused import here

spearec commented on src/client/app/actions/groups.ts:
```diff
@@ -77,290 +97,129 @@ export function fetchGroupChildrenIfNeeded(groupID: number) {
-			const rawGroup = getState().groups.groupInEditing as t.GroupDefinition;
-			if (rawGroup === undefined) {
-				throw new Error('Unacceptable condition: state.groups.groupInEditing has no data.');
+export function submitGroupEdits(group: t.GroupEditData, reload: boolean = true): Thunk {
```
This function contains a fair bit of todo/unused code, not sure if it needs to be cleaned up or if this is marked as part of the issue with reloading pages and will be fixed at a later date.

huss replied:
Thanks for the reminder. As I feared, it took a little work to make sure state was properly used to update pages if a reload is not done but that is all fixed now. Thus, it only reloads when needed and dispatch is again used.

spearec commented on src/client/app/translations/data.js:
```diff
@@ -132,30 +131,52 @@ const localeData = {
 		"group.all.meters": "All Meters",
-		"group.gps": "Group GPS (latitude, longitude)",
+		"group.area": "Group Area:",
+		"group.children.error": "You cannot have a group that has no child meters nor groups.",
```
Maybe rephrase to "Groups must have at least one child meter or group."?

huss replied:
I agree so changed.

spearec commented on src/client/app/translations/data.js:
```diff
@@ -132,30 +131,52 @@ const localeData = {
+		"group.edit.cancelled": "THE CHANGE TO THE GROUP IS CANCELLED",
+		"group.edit.changed": "will have its compatible units changed by the edit to this group",
+		"group.edit.circular": "Adding this meter/group to this group creates a circular dependency.",
+		"group.edit.empty": "Removing this meter/group means there are no meters nor groups which is not allowed . Delete the group if you want to remove it.",
```
add "child"

spearec replied:
this is in reference to line 155

spearec commented on src/client/app/translations/data.js:
```diff
@@ -132,30 +131,52 @@ const localeData = {
+		"group.edit.circular": "Adding this meter/group to this group creates a circular dependency.",
+		"group.edit.empty": "Removing this meter/group means there are no meters nor groups which is not allowed . Delete the group if you want to remove it.",
+		"group.edit.nocompatible": "would have no compatible units by the edit to this group so the edit is cancelled",
+		"group.edit.nounit": "will have its compatible units changed and its default graphic unit set to no unit by the edit to this group",
```
consider replacing with "no unit" (adding quotes)

spearec replied:
this is in reference to line 157

spearec commented:
When attempting to delete a group which is a child of other groups, the improper error is given:

![image](https://user-images.githubusercontent.com/113216142/233730171-18f0f3b5-09cf-4221-ac31-a9ecb8a3eac7.png)

Should say "is contained in" or something similar.

spearec reviewed (changes requested):
Overall, code looks good. I have a few comments, but nothing too major.
I tried to break the groups edit/create pages, and most of the problems I found earlier have been fixed. The UI is far more intuitive than the old page, and the error messages are informative.
Having also already tested this in my area normalization branch, I think that once these comments are addressed this should be good to go.

huss committed:
fix translations per review

huss committed:
review comment fixes & issues

- Removed commented out import.
- Only reload when needed.
- Make create/edit update if another group is edited so menus update.
- Fix bug where save put own group two times. Not see before because
web browser refreshed.
- Fix bug so group names update if another group is edited.

huss committed:
fix bad message noted in PR comment

huss commented:
> When attempting to delete a group which is a child of other groups, the improper error is given:
> 
> ![image](https://user-images.githubusercontent.com/113216142/233730171-18f0f3b5-09cf-4221-ac31-a9ecb8a3eac7.png)
> 
> Should say "is contained in" or something similar.

Thanks. It is now fixed.

huss commented:
@spearec reviewed the code and I am accepting that for merging. All comments were addressed. It is time to merge this big change.
